src/libstd/collections/hash/map.rs

```diff
@@ -353,37 +352,50 @@ fn search_hashed<K, V, M, F>(table: M,
     // undefined behavior when Bucket::new gets the raw bucket in this
     // case, immediately return the appropriate search result.
     if table.capacity() == 0 {
-        return TableRef(table);
+        return InternalEntry::TableIsEmpty;
```

Comment:
We have two groups of methods: inserting (insert, entry, Recover::replace) which reserve space for an element, and other accessors (get, contains_key, remove, Recover::take) which have no effect on an empty table. All of the latter handle `TableIsEmpty` by returning `None`/`false`.

Are you suggesting to check for `capacity() == 0` twice, first in non-inserting accessors and later in `search_hashed`?